New locale: Bokmål Norwegian

Hello,

I would like a Norwegian locale.

If there is some more information needed, please let me know (sorry if this is too barebones :) ).

I'm in the `ebook-reader-dict`. 

```
______________________ ERROR collecting tests/tests_no.py ______________________
ImportError while importing test module '/ebook-reader-dict-no/tests/tests_no.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/local/Cellar/python@3.9/3.9.4/Frameworks/Python.framework/Versions/3.9/lib/python3.9/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
/ebook-reader-dict-ca/tests/tests_no.py:3: in <module>
    ???
E   ModuleNotFoundError: No module named 'wikidict.render'
```
It also links to ebook-reader-dict-ca, which I don't really understand (I used Catalonian lang). Might that be something I am missing?

--- a/.github/workflows/auto-updates.yml
+++ b/.github/workflows/auto-updates.yml
@@ -47,6 +47,7 @@ jobs:
           - sv
           - it
           - ca
+          - no
 
     steps:
 

--- a/README.md
+++ b/README.md
@@ -14,6 +14,7 @@ Finally decent dictionaries based on [Wiktionary](https://www.wiktionary.org/) f
 - [Español](https://github.com/BoboTiG/ebook-reader-dict/releases/tag/es)
 - [Français](https://github.com/BoboTiG/ebook-reader-dict/releases/tag/fr)
 - [Italiano](https://github.com/BoboTiG/ebook-reader-dict/releases/tag/it)
+- [norsk](https://github.com/BoboTiG/ebook-reader-dict/releases/tag/no)
 - [Português](https://github.com/BoboTiG/ebook-reader-dict/releases/tag/pt)
 - [Svenska](https://github.com/BoboTiG/ebook-reader-dict/releases/tag/sv)
 

--- a/tests/data/no/konsentrasjon.wiki
+++ b/tests/data/no/konsentrasjon.wiki
@@ -1,0 +1,41 @@
+==Norsk==
+===Substantiv===
+{{no-sub|m}}
+
+#Det å [[konsentrere]] seg; ha stort fokus på noe.
+#:'''''Konsentrasjon''' er viktig for å ikke bli avledet.''
+#{{kjemi|no}} Andelen stoff i noe; mengde stoff løst pr. enhet.
+#:'''''Konsentrasjonen''' i løsningen er på 0,1 [[molar]].''
+
+====Etymologi====
+Fra {{suffiks|konsentrere|sjon|språk=no}}
+
+====Grammatikk====
+{{no-sub-m1}}
+
+====Oversettelser====
+{{overs-topp|fokusere}}
+* {{overs|da|koncentration|c}}
+* {{overs|en|concentration}}
+* {{overs|fi|keskittyminen}}
+* {{overs|fr|concentration|f}}
+* {{overs-mangler|it}}
+{{overs-midt}}
+* {{overs-mangler|ru}}
+* {{overs-mangler|es}}
+* {{overs|sv|koncentration|c}}
+* {{overs|de|Konzentration|f}}
+{{overs-bunn}}
+{{overs-topp|begrep innen kjemi}}
+* {{overs|da|koncentration|c}}
+* {{overs|en|concentration}}
+* {{overs|fi|konsentraatio}}
+* {{overs|fr|concentration|f}}
+* {{overs-mangler|it}}
+{{overs-midt}}
+* {{overs-mangler|ru}}
+* {{overs-mangler|es}}
+* {{overs|sv|koncentration|c}}
+* {{overs|de|Konzentration|f}}
+{{overs-bunn}}
+

--- a/tests/test_no.py
+++ b/tests/test_no.py
@@ -1,0 +1,40 @@
+import pytest
+
+from wikidict.render import parse_word
+from wikidict.utils import process_templates
+
+
+@pytest.mark.parametrize(
+    "word, pronunciations, genre, etymology, definitions",
+    [
+        (
+            "konsentrasjon",
+            [],
+            "",
+            ["Fra"],
+            [
+                "Det å konsentrere seg; ha stort fokus på noe.",
+                "<i>(Kjemi)</i> Andelen stoff i noe; mengde stoff løst pr. enhet.",
+            ],
+        ),
+    ],
+)
+def test_parse_word(word, pronunciations, genre, etymology, definitions, page):
+    """Test the sections finder and definitions getter."""
+    code = page(word, "no")
+    details = parse_word(word, code, "no", force=True)
+    assert pronunciations == details.pronunciations
+    assert genre == details.genre
+    assert etymology == details.etymology
+    assert definitions == details.definitions
+
+
+@pytest.mark.parametrize(
+    "wikicode, expected",
+    [
+        ("", ""),
+    ],
+)
+def test_process_templates(wikicode, expected):
+    """Test templates handling."""
+    assert process_templates("foo", wikicode, "no") == expected

--- a/wikidict/lang/__init__.py
+++ b/wikidict/lang/__init__.py
@@ -1,7 +1,7 @@
 """Internationalization stuff."""
 import re
 
-from . import ca, defaults, en, es, fr, it, pt, sv
+from . import ca, defaults, en, es, fr, it, no, pt, sv
 from .ca.langs import langs as CA
 from .en.langs import langs as EN
 from .es.langs import langs as ES
@@ -35,6 +35,7 @@ genre = {
     "es": re.compile(defaults.genre),
     "fr": re.compile(fr.genre),
     "it": re.compile(defaults.genre),
+    "no": re.compile(defaults.genre),
     "pt": re.compile(pt.genre),
     "sv": re.compile(defaults.genre),
 }
@@ -46,6 +47,7 @@ float_separator = {
     "es": es.float_separator,
     "fr": fr.float_separator,
     "it": it.float_separator,
+    "no": no.float_separator,
     "pt": pt.float_separator,
     "sv": sv.float_separator,
 }
@@ -57,6 +59,7 @@ thousands_separator = {
     "es": es.thousands_separator,
     "fr": fr.thousands_separator,
     "it": it.thousands_separator,
+    "no": no.thousands_separator,
     "pt": pt.thousands_separator,
     "sv": sv.thousands_separator,
 }
@@ -68,6 +71,7 @@ section_patterns = {
     "es": defaults.section_patterns,
     "fr": defaults.section_patterns,
     "it": defaults.section_patterns,
+    "no": defaults.section_patterns,
     "pt": pt.section_patterns,
     "sv": defaults.section_patterns,
 }
@@ -77,6 +81,7 @@ sublist_patterns = {
     "es": defaults.sublist_patterns,
     "fr": defaults.sublist_patterns,
     "it": defaults.sublist_patterns,
+    "no": defaults.sublist_patterns,
     "pt": defaults.sublist_patterns,
     "sv": defaults.sublist_patterns,
 }
@@ -86,6 +91,7 @@ section_level = {
     "es": defaults.section_level,
     "fr": defaults.section_level,
     "it": defaults.section_level,
+    "no": defaults.section_level,
     "pt": pt.section_level,
     "sv": defaults.section_level,
 }
@@ -95,6 +101,7 @@ section_sublevels = {
     "es": es.section_sublevels,
     "fr": defaults.section_sublevels,
     "it": defaults.section_sublevels,
+    "no": no.section_sublevels,
     "pt": pt.section_sublevels,
     "sv": defaults.section_sublevels,
 }
@@ -104,6 +111,7 @@ head_sections = {
     "es": es.head_sections,
     "fr": fr.head_sections,
     "it": it.head_sections,
+    "no": no.head_sections,
     "pt": pt.head_sections,
     "sv": sv.head_sections,
 }
@@ -113,6 +121,7 @@ etyl_section = {
     "es": es.etyl_section,
     "fr": fr.etyl_section,
     "it": it.etyl_section,
+    "no": no.etyl_section,
     "pt": pt.etyl_section,
     "sv": "",
 }
@@ -122,6 +131,7 @@ sections = {
     "es": es.sections,
     "fr": fr.sections,
     "it": it.sections,
+    "no": no.sections,
     "pt": pt.sections,
     "sv": sv.sections,
 }
@@ -133,6 +143,7 @@ definitions_to_ignore = {
     "es": es.definitions_to_ignore,
     "fr": fr.definitions_to_ignore,
     "it": it.definitions_to_ignore,
+    "no": defaults.definitions_to_ignore,
     "pt": pt.definitions_to_ignore,
     "sv": defaults.definitions_to_ignore,
 }
@@ -144,6 +155,7 @@ words_to_keep = {
     "es": defaults.words_to_keep,
     "fr": fr.words_to_keep,
     "it": defaults.words_to_keep,
+    "no": defaults.words_to_keep,
     "pt": defaults.words_to_keep,
     "sv": defaults.words_to_keep,
 }
@@ -165,6 +177,7 @@ templates_ignored = {
     "es": es.templates_ignored,
     "fr": fr.templates_ignored,
     "it": it.templates_ignored,
+    "no": defaults.templates_ignored,
     "pt": pt.templates_ignored,
     "sv": sv.templates_ignored,
 }
@@ -180,6 +193,7 @@ templates_italic = {
     "es": es.templates_italic,
     "fr": fr.templates_italic,
     "it": it.templates_italic,
+    "no": defaults.templates_italic,
     "pt": pt.templates_italic,
     "sv": defaults.templates_italic,
 }
@@ -206,6 +220,7 @@ templates_multi = {
     "es": es.templates_multi,
     "fr": fr.templates_multi,
     "it": it.templates_multi,
+    "no": no.templates_multi,
     "pt": pt.templates_multi,
     "sv": sv.templates_multi,
 }
@@ -217,6 +232,7 @@ templates_other = {
     "es": defaults.templates_other,
     "fr": fr.templates_other,
     "it": defaults.templates_other,
+    "no": defaults.templates_other,
     "pt": defaults.templates_other,
     "sv": defaults.templates_other,
 }
@@ -229,6 +245,7 @@ last_template_handler = {
     "es": es.last_template_handler,
     "fr": fr.last_template_handler,
     "it": defaults.last_template_handler,
+    "no": defaults.last_template_handler,
     "pt": pt.last_template_handler,
     "sv": sv.last_template_handler,
 }
@@ -241,6 +258,7 @@ release_description = {
     "es": es.release_description,
     "fr": fr.release_description,
     "it": it.release_description,
+    "no": no.release_description,
     "pt": pt.release_description,
     "sv": sv.release_description,
 }
@@ -252,6 +270,7 @@ wiktionary = {
     "es": es.wiktionary,
     "fr": fr.wiktionary,
     "it": it.wiktionary,
+    "no": no.wiktionary,
     "pt": pt.wiktionary,
     "sv": sv.wiktionary,
 }

--- a/wikidict/lang/no/__init__.py
+++ b/wikidict/lang/no/__init__.py
@@ -1,0 +1,45 @@
+"""Norwegian language."""
+
+# Regex to find the pronunciation
+pronunciation = r""
+
+# Float number separator
+float_separator = ","
+
+# Thousads separator
+thousands_separator = " "
+
+# Markers for sections that contain interesting text to analyse.
+head_sections = ("norsk",)
+section_sublevels = (3, 4)
+etyl_section = ["Etymologi"]
+sections = (
+    *etyl_section,
+    "Substantiv",
+)
+
+# Templates to ignore: the text will be deleted.
+templates_ignored = ("",)
+
+# Templates more complex to manage.
+templates_multi = {
+    "": "",
+}
+
+# Release content on GitHub
+# https://github.com/BoboTiG/ebook-reader-dict/releases/tag/no
+release_description = """\
+Ord räknas: {words_count}
+Dumpa Wiktionary: {dump_date}
+
+Tillgängliga filer:
+
+- [Kobo]({url_kobo}) (dicthtml-{locale}.zip)
+- [StarDict]({url_stardict}) (dict-{locale}.zip)
+- [DictFile]({url_dictfile}) (dict-{locale}.df)
+
+<sub>Uppdaterad på {creation_date}</sub>
+"""  # noqa
+
+# Dictionary name that will be printed below each definition
+wiktionary = "Wiktionary (ɔ) {year}"

--- a/wikidict/render.py
+++ b/wikidict/render.py
@@ -128,7 +128,7 @@ def find_etymology(
     definitions: List[Definitions] = []
     etyl: str
 
-    if locale == "ca":
+    if locale in ("ca", "no"):
         definitions.append(
             process_templates(word, clean(parsed_section.contents), locale)
         )
@@ -204,8 +204,10 @@ def find_genre(code: str, pattern: Pattern[str]) -> str:
     return groups[0] or ""
 
 
-def find_pronunciations(code: str, pattern: Pattern[str]) -> List[str]:
+def find_pronunciations(code: str, pattern: Optional[Pattern[str]]) -> List[str]:
     """Find pronunciations."""
+    if not pattern:
+        return []
     match = pattern.search(code)
     if not match:
         return []
@@ -302,7 +304,7 @@ def parse_word(word: str, code: str, locale: str, force: bool = False) -> Word:
     definitions = find_definitions(word, parsed_sections, locale)
 
     if definitions or force:
-        prons = find_pronunciations(code, pronunciation[locale])
+        prons = find_pronunciations(code, pronunciation.get(locale))
         nature = find_genre(code, genre[locale])
 
     # find if variant and delete unwanted definitions

